fbpic/lpa_utils/laser/antenna_injection.py

```diff
@@ -289,272 +327,101 @@ def deposit( self, fld, fieldtype ):
-        Add the small-size arrays Jr_buffer, Jt_buffer, Jz_buffer into
-        the full-size arrays Jr, Jt, Jz on the GPU
-
-        Parameters:
-        -----------
-        iz_min: int
+        
+                dim_grid_1d, dim_block_1d = cuda_tpb_bpg_1d( self.Ntot )
+                deposit_J_gpu_unsorted[
+                    dim_grid_1d, dim_block_1d](
+                    x, y, self.d_baseline_z, self.w, q,
+                    vx, vy, self.d_vz,
+                    grid[m].invdz, grid[m].zmin, grid[m].Nz,
+                    grid[m].invdr, grid[m].rmin, grid[m].Nr,
+                    grid[m].Jr, grid[m].Jt, grid[m].Jz,
+                    m, grid[m].d_ruyten_linear_coef)
+
+    def deposit_virtual_particles_cpu( self, q, fieldtype, grid, fld ):
+        x = self.baseline_x + q*self.excursion_x
+        y = self.baseline_y + q*self.excursion_y
 
-        Jr_buffer, Jt_buffer, Jz_buffer: 3darrays of complexs
-            Arrays of shape (Nm, 2, Nr) that store the values of rho
-            in the 2 cells that surround the antenna (for each mode).
+        # Divide particles in chunks (each chunk is handled by a different
+        # thread) and register the indices that bound each chunks
+        ptcl_chunk_indices = get_chunk_indices(self.Ntot, nthreads)
 
-        Jr, Jt, Jz: 2darrays of complexs
-            Arrays of shape (Nz, Nr) that contain rho in the mode m
+        # The set of Ruyten shape coefficients to use for higher modes. 
+        # For Nm > 1, the set from mode 1 is used, since all higher modes have the
+        # same coefficients. For Nm == 1, the coefficients from mode 0 are 
+        # passed twice to satisfy the argument types for Numba JIT.
+        if fld.Nm > 1:
+            ruyten_m = 1
+        else: 
+            ruyten_m = 0
 
-        m: int
-           The index of the azimuthal mode involved
-        """
-        # Use one thread per radial cell
-        ir = cuda.grid(1)
+        if fieldtype == 'rho' :
+            # ---------------------------------------
+            # Deposit the charge density all modes at once
+            # ---------------------------------------
+            deposit_rho_numba_linear(
+                x, y, self.baseline_z, self.w, q,
+                grid[0].invdz, grid[0].zmin, grid[0].Nz,
+                grid[0].invdr, grid[0].rmin, grid[0].Nr,
+                fld.rho_global, fld.Nm,
+                nthreads, ptcl_chunk_indices,
+                grid[0].ruyten_linear_coef,
+                grid[ruyten_m].ruyten_linear_coef )
 
-        # Add the values
-        if ir < Jr.shape[1]:
-            Jr[iz_min, ir] += Jr_buffer[m, 0, ir]
-            Jr[iz_min+1, ir] += Jr_buffer[m, 1, ir]
 
-            Jt[iz_min, ir] += Jt_buffer[m, 0, ir]
-            Jt[iz_min+1, ir] += Jt_buffer[m, 1, ir]
+        elif fieldtype == 'J' :
+            # Calculate the relativistic momenta from the velocities
+            # The gamma is set to 1 both here and in the deposition kernel
```

Comment:
Could you add a comment here saying that this works because the current deposition in fact only depends on `ux*inv_gamma`, `uy*inv_gamma` and `uz*inv_gamma`, so as long as we make sure that this product correspond to `vx`, `vy`, `vz`, we can indeed pass `inv_gamma = 1.`.